Fail MSQ compaction if multi-valued partition dimensions are found

gargvishesh opened the pull request:
MSQ currently supports only single-valued string dimensions as partition keys. This patch adds a check to ensure that partition keys are single-valued in case this info is available by virtue of segment download for schema inference. 

During compaction, if MSQ finds multi-valued dimensions (MVDs) declared as part of `range` partitionsSpec, it switches partitioning type to `dynamic`, ending up in repeated compactions of the same interval.  To avoid this scenario, the segment download logic is also updated to always download segments if info on multi-valued dimensions is required.

kfaraz commented:
> the partitioning eventually ends up getting switched to dynamic type.

@gargvishesh , is this something MSQ always does silently?

kfaraz commented on server/src/main/java/org/apache/druid/client/indexing/ClientCompactionRunnerInfo.java:
```diff
@@ -107,7 +108,7 @@ public static CompactionConfigValidationResult validateCompactionConfig(
    * <ul>
    * <li>partitionsSpec of type HashedParititionsSpec.</li>
-   * <li>'range' partitionsSpec with non-string partition dimensions.</li>
+   * <li>'range' partitionsSpec with non-single-valued-string partition dimensions.</li>
```
```suggestion
   * <li>'range' partitionsSpec with multi-valued or non-string partition dimensions.</li>
```

kfaraz commented on server/src/main/java/org/apache/druid/client/indexing/ClientCompactionRunnerInfo.java:
```diff
@@ -138,13 +141,19 @@ private static CompactionConfigValidationResult compactionConfigSupportedByMSQEn
   /**
    * Validate that partitionSpec is either 'dynamic` or 'range'. If 'dynamic', ensure 'maxTotalRows' is null. If range
-   * ensure all partition columns are of string type.
+   * ensure all partition columns are of single-valued string type .
```
```suggestion
   * ensure all partition columns are single-valued and have type string.
```

kfaraz commented on server/src/main/java/org/apache/druid/client/indexing/ClientCompactionRunnerInfo.java:
```diff
@@ -138,13 +141,19 @@ private static CompactionConfigValidationResult compactionConfigSupportedByMSQEn
   {
+    if (partitionsSpec == null) {
+      return CompactionConfigValidationResult.failure(
+          "MSQ: tuningConfig.partitionsSpec must be set"
```
Why do we need this? I don't think specifying `partitionsSpec` is a requirement in native engine?

gargvishesh replied:
In native, a default spec is applied by the index task based on certain field values: https://github.com/apache/druid/blob/c2149d59a7da1ac4cbb02aaa63a52cb326306dc5/indexing-service/src/main/java/org/apache/druid/indexing/common/task/IndexTask.java#L1247

This is not the case with MSQ and the task fails. 

In the web-ui for compaction, partitionsSpec is populated with some value by default.

kfaraz commented on server/src/main/java/org/apache/druid/client/indexing/ClientCompactionRunnerInfo.java:
```diff
@@ -163,17 +172,26 @@ public static CompactionConfigValidationResult validatePartitionsSpecForMSQ(
       );
-      Optional<String> nonStringDimension = ((DimensionRangePartitionsSpec) partitionsSpec)
+      Optional<String> unsupportedDimension = ((DimensionRangePartitionsSpec) partitionsSpec)
           .getPartitionDimensions()
           .stream()
-          .filter(dim -> !ColumnType.STRING.equals(dimensionSchemaMap.get(dim).getColumnType()))
+          .filter(dim -> !ColumnType.STRING.equals(dimensionSchemaMap.get(dim).getColumnType())
+                         || multiValuedDimensions != null && multiValuedDimensions.contains(dim))
```
For clearer code, let's leave the whole check for non string dimension as-is.
Add a separate loop for checking multi-valued dimensions.

gargvishesh replied:
Done.

gargvishesh commented:
@kfaraz: yes, if MVD fields are found, there is just a log that states shard spec is changed to `NumberedShardSpec` type: https://github.com/apache/druid/blob/c2149d59a7da1ac4cbb02aaa63a52cb326306dc5/extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/exec/ControllerImpl.java#L1104

This in fact also means that compaction will keeping continuing in a loop. I was thinking about looking for a solution to this problem post this PR.

kfaraz commented:
> This in fact also means that compaction will keeping continuing in a loop. I was thinking about looking for a solution to this problem post this PR.

Oh, that can be a problem. When can this happen?

gargvishesh commented:
Right. So this case can happen when there is no need to download the segments, ie. user specifies dimensions, metrics, granularity and rollup in the compaction spec itself, but one of the dimensions happens to be MVD.

kfaraz commented:
Can we ensure we download the segments in this case too?
What would be the drawback?

gargvishesh commented:
Drawback was mainly the high cost of downloading segments. But yes I think this would be the only viable way for now.

gargvishesh commented:
@kfaraz: I have now updated the logic to download segments when we need MVD info. The check is also updated to be more fine grained to avoid redundant download.

kfaraz commented on indexing-service/src/main/java/org/apache/druid/indexing/common/task/CompactionTask.java:
```diff
@@ -452,6 +454,66 @@ public boolean isPerfectRollup()
+   * therefore is particularly for MSQCompactionRunner; it returns true when dimension types aren't known
+   * from CompactionTask spec, or if either rollup or partition dimensions contain any string-type column.
+   */
+  public static boolean needMultiValuedDimensions(CompactionTask compactionTask)
```
No point making this method static if it requires passing the `CompactionTask` into it.
Also, instead of making this method public or package private, `CompactionTask` should instead expose a method `needsToDownloadSegments()` or something. That can internally call this one.

gargvishesh replied:
> No point making this method static if it requires passing the CompactionTask into it.

Agree. One possibility is that its value can be assigned to a final field and the getter is exposed

> Also, instead of making this method public or package private, CompactionTask should instead expose a method needsToDownloadSegments() or something. That can internally call this one.

This would need some refactoring since the effective granularitySpec on which download depends is computed inside `createDataSchemaForIntervals`, just before passing this computed value to `ExistingSegmentAnalyzer` where the final need to download is evaluated. Also, each `need*` flag is separately communicated to the Analyzer class to do selective processing. 

Does the proposed approach of assigning `needMultiValuedDimensions` to a private field and creating tests on it work? In fact, all such fields apart from `needGranularitySpec` can be populated in CompactionTask constructor and passed to the Analyzer

gargvishesh replied:
Just relooked at code. Only segmentGranularity is computed later, so all need* flags and an additional new needsDownload flag can be computed in the CompactionTask constructor itself. The needsDownload can have a public getter for tests.

kfaraz replied:
If only needed in test, keep the method as package private rather than public.

gargvishesh replied:
public is needed since the tests reside in the MSQ extension

kfaraz commented on indexing-service/src/main/java/org/apache/druid/indexing/common/task/CompactionTask.java:
```diff
@@ -452,6 +454,45 @@ public boolean isPerfectRollup()
+                                 && tuningConfig.getPartitionsSpec() instanceof DimensionRangePartitionsSpec;
+
+    if (dimensionsSpec == null || dimensionsSpec.getDimensions().isEmpty()) {
+      return (isPossiblyRollup || isRangePartitioned);
```
```suggestion
      return isPossiblyRollup || isRangePartitioned;
```

kfaraz commented on indexing-service/src/main/java/org/apache/druid/indexing/common/task/CompactionTask.java:
```diff
@@ -452,6 +454,45 @@ public boolean isPerfectRollup()
+   * therefore is particularly for MSQCompactionRunner; it returns true when dimension types aren't known
+   * from CompactionTask spec, or if either rollup or partition dimensions contain any string-type column.
+   */
+  boolean needMultiValuedDimensions()
```
Can this be renamed to something like `needToIdentifyMultiValuedDimensions` or `shouldAnalyzeMultiValuedDimensions` (to correspond to segment analyzer)?
Just `needMvd` seems ambiguous.

gargvishesh replied:
Updated to `identifyMultiValuedDimensions`

kfaraz commented on indexing-service/src/main/java/org/apache/druid/indexing/common/task/CompactionTask.java:
```diff
@@ -452,6 +454,45 @@ public boolean isPerfectRollup()
   }
 
+  /**
+   * In MSQ, MVDs need conversion to array for rollup, and they aren't supported as partition columns. This check
+   * therefore is particularly for MSQCompactionRunner; it returns true when dimension types aren't known
+   * from CompactionTask spec, or if either rollup or partition dimensions contain any string-type column.
```
```suggestion
   * Checks if multi-valued string dimensions need to be analyzed by downloading the segments.
   * This method returns true only for MSQ engine when either of the following holds true:
   * <ul>
   * <li> Range partitioning is done on a string dimension or an unknown dimension
   * (since MSQ does not support partitioning on a multi-valued string dimension) </li>
   * <li> Rollup is done on a string dimension or an unknown dimension
   * (since MSQ requires multi-valued string dimensions to be converted to arrays for rollup) </li>
   * </ul>
   * @return false for native engine, true for MSQ engine only when partitioning or rollup is done on a string or unknown dimension.
```

kfaraz commented on indexing-service/src/main/java/org/apache/druid/indexing/common/task/CompactionTask.java:
```diff
@@ -452,6 +454,45 @@ public boolean isPerfectRollup()
+      return false;
+    }
+    // Rollup can be true even when granularitySpec is not known since rollup is then decided based on segment analysis
+    boolean isPossiblyRollup = !(granularitySpec != null && Boolean.FALSE.equals(granularitySpec.isRollup()));
```
Trying to remove the double negative:
```suggestion
    final boolean isPossiblyRollup = granularitySpec == null || !Boolean.FALSE.equals(granularitySpec.isRollup());
```

kfaraz commented on indexing-service/src/main/java/org/apache/druid/indexing/common/task/CompactionTask.java:
```diff
@@ -452,6 +454,45 @@ public boolean isPerfectRollup()
+      boolean isRollupOnStringDimension = isPossiblyRollup &&
+                                          dimensionsSpec.getDimensions()
+                                                        .stream()
+                                                        .anyMatch(dim -> ColumnType.STRING.equals(
+                                                            dim.getColumnType()));
```
formatting:
```suggestion
                                                        .anyMatch(dim -> ColumnType.STRING.equals(dim.getColumnType()));
```

kfaraz commented on indexing-service/src/main/java/org/apache/druid/indexing/common/task/CompactionTask.java:
```diff
@@ -452,6 +454,45 @@ public boolean isPerfectRollup()
+          dimensionsSpec.getDimensions()
+                        .stream()
+                        .anyMatch(
+                            dim -> ColumnType.STRING.equals(
+                                dim.getColumnType())
```
formatting:
```suggestion
                            dim -> ColumnType.STRING.equals(dim.getColumnType())
```

kfaraz commented on indexing-service/src/test/java/org/apache/druid/indexing/common/task/CompactionTaskTest.java:
```diff
@@ -2015,6 +2100,37 @@ public SettableColumnValueSelector<?> makeNewSettableColumnValueSelector()
   }
 
+  /**
+   * A class to mimic MSQCompactionRunner behaviour, since the actual class resides in the MSQ extn
```
How are we mimicking the behaviour? The methods seem to return only null.

gargvishesh replied:
Just mimics for validations. Updated the comment

kfaraz commented on extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/MSQCompactionRunner.java:
```diff
@@ -175,6 +187,31 @@ public CompactionConfigValidationResult validateCompactionTask(
                             .orElse(CompactionConfigValidationResult.success());
   }
 
+  private CompactionConfigValidationResult validatePartitionDimensionsNotMultivalued(
```
```suggestion
  private CompactionConfigValidationResult validatePartitionDimensionsAreNotMultiValued(
```

kfaraz commented on extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/MSQCompactionRunner.java:
```diff
@@ -175,6 +187,31 @@ public CompactionConfigValidationResult validateCompactionTask(
+    List<String> dimensionSchemas = dimensionsSpec.getDimensionNames();
+    if (partitionsSpec instanceof DimensionRangePartitionsSpec
+        && dimensionSchemas != null
+        && multiValuedDimensions != null) {
```
Check for `multiValuedDimensions` empty too?

gargvishesh replied:
Yes makes it a bit more efficient. Thanks

kfaraz commented on extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/MSQCompactionRunner.java:
```diff
@@ -153,13 +155,23 @@ public CompactionConfigValidationResult validateCompactionTask(
+              dataSchema instanceof CombinedDataSchema
+              ? ((CombinedDataSchema) dataSchema).getMultiValuedDimensions()
+              : null
+          ));
```
format:
```suggestion
          )
      );
```

kfaraz reviewed (approved):
Approach looks good. Minor cleanup suggestions.
